### PR TITLE
ExampleMod Solar Eruption Item PR

### PR DESCRIPTION
**Pull request by Pyroknight**

### CODE REVIEW

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+			player.ChangeDir(projectile.direction);
+			player.heldProj = projectile.whoAmI;
+			player.itemTime = 2;
+			player.itemAnimation = 2;
+			player.itemRotation = (projectile.velocity * projectile.direction).ToRotation();
+
+			// Makes some dust and light.
+			for (int i = 0; i < 3; i++)
+			{
+				Dust dust1 = Dust.NewDustPerfect(chainHeadPosition, DustID.Smoke, new Vector2(0f, -1f).RotatedByRandom(MathHelper.ToRadians(360f)), 150, Color.White, 1.5f);
```
Comment (RighteousRyan1) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:54]: As with this, maybe change something like `dust1` to something... different?

Comment (RighteousRyan1) [reply ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:54]: Same with the others

File: ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		{
+			projectile.width = 52;
+			projectile.height = 52;
+			projectile.friendly = true;
+			projectile.alpha = 255;
+			projectile.ignoreWater = true;
+			projectile.timeLeft = 60;
+			projectile.tileCollide = false;
+			projectile.penetrate = -1;
+			projectile.usesLocalNPCImmunity = true;
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs:28]: I feel like stuff such as this should have comments explaining what they actually do.

File: ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			// Add light in place of the explosion. Color.White can be any color you want, and 1f at the end is a radius multiplier.
+			//Lighting.AddLight(projectile.Center, Color.White.ToVector3() * 1f);
+			
+			// Make the explosion larger overtime. By the time of death (60 ticks) it grows by 0.6f.
+			projectile.ai[1] += 0.01f;
+			projectile.scale = projectile.ai[1];
+
+			// When it spawns, play an explosion sound and 
+			if (projectile.ai[0] == 0)
+            {
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs:43]: Formatting :trollface:

File: ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			projectile.Damage();
+
+			// Basic explosion dust
+			int dusts = 5;
+			for (int i = 0; i < dusts; i++)
+			{
+				if (Main.rand.NextBool(3))
+				{
+					float speed = 6f;
+					Dust dust1 = Dust.NewDustPerfect(projectile.Center, DustID.Smoke, new Vector2(0f, -speed * Main.rand.NextFloat(0.5f, 1.2f)).RotatedBy(MathHelper.ToRadians(360f / i * dusts + Main.rand.NextFloat(-50f, 50f))), 150, Color.White, 1.5f);
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs:81]: Might help to comment explaining what the vector calculations are

File: ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+					dust1.noGravity = true;
+					dust1.shader = GameShaders.Armor.GetSecondaryShader(player.cPet, player);
+				}
+			}
+		}
+
+		public override void OnHitNPC(NPC target, int damage, float knockback, bool crit)
+		{
+			int cooldown = 4;
+			projectile.localNPCImmunity[target.whoAmI] = 6;
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionExplosion.cs:91]: Comments would help

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+			projectile.Center = player.RotatedRelativePoint(player.position + offset) - projectile.velocity;
+		}
+
+		public override void OnHitNPC(NPC target, int damage, float knockback, bool crit)
+		{
+			// Spawns an explosion when it hits an NPC.
+			int cooldown = 3;
+			if (projectile.localAI[1] <= 0f)
+			{
+				Projectile explosion = Projectile.NewProjectileDirect(target.Center, Vector2.Zero, ModContent.ProjectileType<ExampleSolarEruptionExplosion>(), damage, 10f, projectile.owner, 0f, 1f);
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:108]: I honestly find a lot of this hard to follow
Splitting it up into local vars with comments would likely help a new-comer

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		}
+
+		public override bool? CanCutTiles()
+		{
+			return true;
+		}
+
+		public override void CutTiles()
+		{
+			DelegateMethods.tilecut_0 = TileCuttingContext.AttackProjectile;
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:125]: An explanation for this would be nice

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+			// Custom collision so all chains across the flail can cause impact.
+			float collisionPoint = 0f;
+			if (Collision.CheckAABBvLineCollision(targetHitbox.TopLeft(), targetHitbox.Size(), projectile.Center, projectile.Center + projectile.velocity, (projectile.width + projectile.height) * 0.5f * projectile.scale, ref collisionPoint))
+			{
+				return true;
+			}
+			return false;
+		}
+
+		public override bool PreDraw(SpriteBatch spriteBatch, Color lightColor)
```
Comment (Steviegt6) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:140]: Heavily lacking in comments

**Review by steviegt6 — CHANGES_REQUESTED** [review]

**Review by RighteousRyan1 — COMMENTED** [review]
Generally agreeing with Stevie, comments are needed in lots of spots and code is a bit messy. Maybe fix that up a bit then it should be good in my opinion.

**Review by Pyroknight — COMMENTED** [review]
Finished reviewing changes from Stevie and Ryan, made changes accordingly.

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+            if (chainDistance > 0f)
+            {
+                float chains = 0f;
+                while (chains + 1f < chainDistance)
+                {
+                    if (chainDistance - chains < rectangle.Height)
+                    {
+                        rectangle.Height = (int)(chainDistance - chains);
+                    }
+                    spriteBatch.Draw(texture, position - Main.screenPosition + yOffset, rectangle, Lighting.GetColor((int)position.X / 16, (int)position.Y / 16), rotation, new Vector2(rectangle.Width / 2, 0f), projectile.scale, SpriteEffects.None, 0f);
```
Comment (Mellozx) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:194]: Same as above

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+            {
+                return false;
+            }
+            float chainDistance = projectile.velocity.Length() + 16f;
+            bool distanceCheck = chainDistance < 100f;
+            Vector2 direction = Vector2.Normalize(projectile.velocity);
+            Rectangle rectangle = chainHandle;
+            Vector2 yOffset = new Vector2(0f, Main.player[projectile.owner].gfxOffY);
+            float rotation = projectile.rotation;
+            spriteBatch.Draw(texture, projectile.Center - Main.screenPosition + yOffset, rectangle, color, rotation, rectangle.Size() / 2f - Vector2.UnitY * 4f, projectile.scale, SpriteEffects.None, 0f);
```
Comment (Mellozx) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:180]: A comment around the spritebatches would be nice, to clarify what you are doing

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+                chains += increment * 0.25f;
+                position += direction * increment * 0.25f;
+                for (int i = 0; i < offset; i++)
+                {
+                    float spacing = increment;
+                    if (i == 0)
+                    {
+                        spacing *= 0.75f;
+                    }
+                    spriteBatch.Draw(texture, position - Main.screenPosition + yOffset, rectangle, Lighting.GetColor((int)position.X / 16, (int)position.Y / 16), rotation, new Vector2(rectangle.Width / 2, 0f), projectile.scale, SpriteEffects.None, 0f);
```
Comment (Mellozx) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:218]: Also here

File: ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                    chains += spacing;
+                    position += direction * spacing;
+                }
+            }
+            rectangle = chainHead;
+            spriteBatch.Draw(texture, chainEnd - Main.screenPosition + yOffset, rectangle, Lighting.GetColor((int)chainEnd.X / 16, (int)chainEnd.Y / 16), rotation, texture.Frame().Top(), projectile.scale, SpriteEffects.None, 0f);
```
Comment (Mellozx) [inline ExampleMod/Projectiles/ExampleSolarEruptionProjectile.cs:224]: and here, since there isn't much clarification about what's being drawn

**Review by Mellozx — CHANGES_REQUESTED** [review]

**Review by Pyroknight — COMMENTED** [review]
More fixes and indentation

**Review by Pyroknight — COMMENTED** [review]
Brace Indentation Fixes

**Comment by Mirsario** [comment]
There is not much use in making PRs to 1.3 at this point, especially its ExampleMod. We're rebuilding EM in 1.4, and the quality standard is way higher than 1.3's. You should move this to a new PR targeting `1.4`. When you do that - it's a also good idea to make a branch with an unique name, like `1.4_solareruption`

**Review by Mirsario — CHANGES_REQUESTED** [review]
There is not much use in making PRs to 1.3 at this point, especially its ExampleMod. We're rebuilding EM in 1.4, and the quality standard is way higher than 1.3's. You should move this to a new PR targeting `1.4`. When you do that - it's a also good idea to make a branch with an unique name, like `1.4_solareruption`
